File: 0162-find-peak-element/0162-find-peak-element.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int n = nums.size() ;
        if (n==1) return 0;
        if (nums[0] > nums[1]) return 0;
        if(nums[n-1] > nums[n-2]) return n-1;
        
        int l = 1, h = n-2;
        while(l <= h) {
            int m = (l + h) >> 1;
            if (nums[m] >= nums[m-1] && nums[m] >= nums[m+1]){
                return m;
            } else if ( nums[m] > nums[m-1]) l = m + 1;
            else h = m-1;
        }
        return -1;
    }
};

File: 0162-find-peak-element/0162-find-peak-element.py
class Solution:
    def findPeakElement(self, a: List[int]) -> int:
        
        if len(a) == 1:
            return 0
        if a[0] > a[1]: return 0
        if a[len(a)-1] > a[len(a) - 2]: return len(a)-1
        l, r = 1, len(a) - 2
        while l <= r:
            m = (l + r) >> 1
            if  a[m] > a[m-1] and a[m+1] < a[m]:
                return m

            if a[m] > a[m-1]:
                l = m + 1
            else:
                r = m - 1 
        